File: internal/proxy/proxy.go
package proxy

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync/atomic"
	"time"

	"github.com/SalzDevs/agenttop/internal/event"
	"github.com/SalzDevs/agenttop/internal/pricing"
	"github.com/SalzDevs/agenttop/internal/store"
)

type Proxy struct {
	Store           *store.Store
	Bus             *event.Bus
	Client          *http.Client
	Port            int
	traceID         int64
	AnthropicTarget string
	OpenAITarget    string
}

func New(s *store.Store, b *event.Bus, port int) *Proxy {
	return &Proxy{
		Store:           s,
		Bus:             b,
		Port:            port,
		Client:          &http.Client{Timeout: 0},
		AnthropicTarget: "https://api.anthropic.com",
		OpenAITarget:    "https://api.openai.com",
	}
}

// NewWithUpstreams is like New but lets the caller override the upstream
// provider URLs. Useful for local LLMs, corporate proxies, or testing.
func NewWithUpstreams(s *store.Store, b *event.Bus, port int, anthropicURL, openaiURL string) *Proxy {
	p := New(s, b, port)
	if anthropicURL != "" {
		p.AnthropicTarget = anthropicURL
	}
	if openaiURL != "" {
		p.OpenAITarget = openaiURL
	}
	return p
}

type route struct {
	provider string
	target   string
}

func (p *Proxy) route(r *http.Request) route {
	// x-agenttop-upstream: explicit upstream override (used by opencode-go,
	// opencode-zen, and any custom provider that sets this header). This lets
	// the proxy forward to the real API even for providers it doesn't know about.
	if upstream := r.Header.Get("x-agenttop-upstream"); upstream != "" {
		// Determine provider format from the path/header for parsing purposes.
		provider := "openai"
		if r.Header.Get("anthropic-version") != "" || strings.HasPrefix(r.URL.Path, "/v1/messages") {
			provider = "anthropic"
		}
		return route{provider, upstream}
	}
	if h := r.Header.Get("x-agenttop-provider"); h != "" {
		if h == "anthropic" {
			return route{"anthropic", p.AnthropicTarget}
		}
		return route{"openai", p.OpenAITarget}
	}
	// The Anthropic SDK (used by opencode, Claude Code, etc.) sends
	// `anthropic-version` on every request. This is more robust than
	// path matching: it also catches /v1/messages/count_tokens and any
	// future endpoints, routing them to Anthropic instead of OpenAI.
	if r.Header.Get("anthropic-version") != "" {
		return route{"anthropic", p.AnthropicTarget}
	}
	if strings.HasPrefix(r.URL.Path, "/v1/messages") {
		return route{"anthropic", p.AnthropicTarget}
	}
	return route{"openai", p.OpenAITarget}
}

func (p *Proxy) emit(e event.Event) {
	e = p.Store.Append(e)
	p.Bus.Emit(e)
}

func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	rt := p.route(r)
	traceID := atomic.AddInt64(&p.traceID, 1)
	start := time.Now()

	body, err := io.ReadAll(r.Body)
	_ = r.Body.Close()
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}

	meta := parseRequest(rt.provider, body, r)

	startEvt := event.Event{
		TraceID:       traceID,
		Time:          start,
		Provider:      rt.provider,
		Model:         meta.model,
		Endpoint:      r.URL.Path,
		Method:        r.Method,
		Streaming:     meta.streaming,
		PromptPreview: meta.promptPreview,
	}
	p.emit(startEvt)

	outURL := rt.target + r.URL.Path
	if r.URL.RawQuery != "" {
		outURL += "?" + r.URL.RawQuery
	}
	outReq, err := http.NewRequestWithContext(r.Context(), r.Method, outURL, bytes.NewReader(body))
	if err != nil {
		p.emit(fail(startEvt, start, err.Error()))
		http.Error(w, err.Error(), 502)
		return
	}
	copyHeaders(outReq.Header, r.Header)
	outReq.Header.Del("x-agenttop-provider")
	outReq.Header.Del("x-agenttop-upstream")

	resp, err := p.Client.Do(outReq)
	if err != nil {
		p.emit(fail(startEvt, start, err.Error()))
		http.Error(w, err.Error(), 502)
		return
	}

	hdr := w.Header()
	copyHeaders(hdr, resp.Header)
	hdr.Del("Content-Length")
	w.WriteHeader(resp.StatusCode)

	capture := newCaptureWriter()
	flusher, _ := w.(http.Flusher)
	buf := make([]byte, 4096)
	for {
		n, rerr := resp.Body.Read(buf)
		if n > 0 {
			w.Write(buf[:n])
			capture.Write(buf[:n])
			if flusher != nil {
				flusher.Flush()
			}
		}
		if rerr != nil {
			break
		}
	}
	resp.Body.Close()

	usage, respPreview := parseResponse(rt.provider, resp.Header, capture.Bytes(), meta)

	endEvt := startEvt
	endEvt.Time = time.Now()
	endEvt.Status = resp.StatusCode
	endEvt.Duration = time.Since(start)
	endEvt.InputTokens = usage.in
	endEvt.OutputTokens = usage.out
	endEvt.CacheReadTokens = usage.cacheRead
	endEvt.CacheWriteTokens = usage.cacheWrite
	endEvt.ResponsePreview = respPreview
	endEvt.CostUSD = pricing.Cost(meta.model, usage.in, usage.out, usage.cacheRead, usage.cacheWrite)
	p.emit(endEvt)
}

func fail(e event.Event, start time.Time, msg string) event.Event {
	e.Time = time.Now()
	e.Status = 0
	e.Duration = time.Since(start)
	e.Err = msg
	return e
}

func copyHeaders(dst, src http.Header) {
	for k, v := range src {
		if strings.EqualFold(k, "Host") || strings.EqualFold(k, "Content-Length") {
			continue
		}
		dst[k] = v
	}
}

func hostOnly(u string) string {
	u = strings.TrimPrefix(u, "https://")
	u = strings.TrimPrefix(u, "http://")
	if i := strings.IndexByte(u, '/'); i >= 0 {
		u = u[:i]
	}
	return u
}

type captureWriter struct {
	buf   []byte
	maxSz int
}

func newCaptureWriter() *captureWriter {
	return &captureWriter{maxSz: 1 << 20} // 1MB — plenty for the final usage chunk
}

func (c *captureWriter) Write(b []byte) (int, error) {
	n := len(b)
	if len(c.buf)+n <= c.maxSz {
		c.buf = append(c.buf, b...)
	} else {
		// Keep the last maxSz bytes (ring buffer) so the final usage SSE
		// chunk — which arrives at the end of the stream — is always captured.
		overflow := len(c.buf) + n - c.maxSz
		if overflow > 0 {
			if overflow >= len(c.buf) {
				c.buf = c.buf[:0]
			} else {
				c.buf = c.buf[overflow:]
			}
		}
		remaining := c.maxSz - len(c.buf)
		if remaining > n {
			remaining = n
		}
		c.buf = append(c.buf, b[n-remaining:]...)
	}
	return n, nil
}

func (c *captureWriter) Bytes() []byte { return c.buf }

type reqMeta struct {
	model         string
	streaming     bool
	promptPreview string
}

type rawMessage struct {
	Role    string          `json:"role"`
	Content json.RawMessage `json:"content"`
}

type reqBody struct {
	Model    string       `json:"model"`
	Stream   bool         `json:"stream"`
	Messages []rawMessage `json:"messages"`
}

func parseRequest(provider string, body []byte, r *http.Request) reqMeta {
	var m reqBody
	if err := json.Unmarshal(body, &m); err != nil {
		return reqMeta{}
	}
	preview := ""
	for i := len(m.Messages) - 1; i >= 0; i-- {
		if strings.EqualFold(m.Messages[i].Role, "user") {
			preview = extractText(m.Messages[i].Content)
			break
		}
	}
	return reqMeta{model: m.Model, streaming: m.Stream, promptPreview: trim(preview, 120)}
}

func extractText(content json.RawMessage) string {
	var s string
	if json.Unmarshal(content, &s) == nil {
		return s
	}
	var parts []struct {
		Type string `json:"type"`
		Text string `json:"text"`
	}
	if json.Unmarshal(content, &parts) == nil {
		var b strings.Builder
		for _, p := range parts {
			if p.Type == "text" || p.Type == "" {
				b.WriteString(p.Text)
			}
		}
		return b.String()
	}
	return ""
}

type usage struct {
	in, out, cacheRead, cacheWrite int
}

func parseResponse(provider string, hdr http.Header, body []byte, meta reqMeta) (usage, string) {
	ct := hdr.Get("Content-Type")
	if strings.Contains(ct, "text/event-stream") {
		return parseSSE(provider, body, meta)
	}
	return parseJSON(provider, body, meta)
}

func parseSSE(provider string, body []byte, meta reqMeta) (usage, string) {
	var u usage
	var sb strings.Builder
	for _, line := range bytes.Split(body, []byte("\n")) {
		line = bytes.TrimSpace(line)
		if !bytes.HasPrefix(line, []byte("data:")) {
			continue
		}
		data := bytes.TrimSpace(bytes.TrimPrefix(line, []byte("data:")))
		if bytes.Equal(data, []byte("[DONE]")) {
			continue
		}
		var raw map[string]json.RawMessage
		if json.Unmarshal(data, &raw) != nil {
			continue
		}
		if provider == "anthropic" {
			if t, ok := raw["type"]; ok {
				var ts string
				json.Unmarshal(t, &ts)
				if ts == "message_start" {
					var wrap struct {
						Message struct {
							Usage struct {
								InputTokens                int `json:"input_tokens"`
								CacheCreationInputTokens   int `json:"cache_creation_input_tokens"`
								CacheReadInputTokens       int `json:"cache_read_input_tokens"`
								OutputTokens               int `json:"output_tokens"`
							} `json:"usage"`
						} `json:"message"`
					}
					json.Unmarshal(data, &wrap)
					u.in = wrap.Message.Usage.InputTokens
					u.cacheWrite = wrap.Message.Usage.CacheCreationInputTokens
					u.cacheRead = wrap.Message.Usage.CacheReadInputTokens
				}
				if ts == "message_delta" {
					var wrap struct {
						Usage struct{ OutputTokens int `json:"output_tokens"` } `json:"usage"`
					}
					json.Unmarshal(data, &wrap)
					if wrap.Usage.OutputTokens > 0 {
						u.out = wrap.Usage.OutputTokens
					}
				}
			}
			extractDeltaText(raw, &sb)
		} else {
			if usageRaw, ok := raw["usage"]; ok && string(usageRaw) != "null" {
				var uu struct {
					PromptTokens         int `json:"prompt_tokens"`
					CompletionTokens     int `json:"completion_tokens"`
					PromptCacheHitTokens int `json:"prompt_cache_hit_tokens"`
				}
				json.Unmarshal(usageRaw, &uu)
				if uu.PromptTokens > 0 {
					u.in = uu.PromptTokens
				}
				if uu.PromptCacheHitTokens > 0 {
					u.cacheRead = uu.PromptCacheHitTokens
				}
				if uu.CompletionTokens > 0 {
					u.out = uu.CompletionTokens
				}
			}
			extractOpenAIDelta(raw, &sb)
		}
	}
	return u, trim(sb.String(), 400)
}

func extractDeltaText(raw map[string]json.RawMessage, sb *strings.Builder) {
	if d, ok := raw["delta"]; ok {
		var dd struct {
			Type string `json:"type"`
			Text string `json:"text"`
		}
		json.Unmarshal(d, &dd)
		if dd.Type == "text_delta" {
			sb.WriteString(dd.Text)
		}
	}
}

func extractOpenAIDelta(raw map[string]json.RawMessage, sb *strings.Builder) {
	if d, ok := raw["choices"]; ok {
		var choices []struct {
			Delta struct {
				Content          string `json:"content"`
				ReasoningContent string `json:"reasoning_content"`
			} `json:"delta"`
		}
		json.Unmarshal(d, &choices)
		for _, c := range choices {
			if c.Delta.Content != "" {
				sb.WriteString(c.Delta.Content)
			}
		}
	}
}

func parseJSON(provider string, body []byte, meta reqMeta) (usage, string) {
	var u usage
	var preview string
	if provider == "anthropic" {
		var r struct {
			Usage struct {
				InputTokens              int `json:"input_tokens"`
				OutputTokens             int `json:"output_tokens"`
				CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
				CacheReadInputTokens     int `json:"cache_read_input_tokens"`
			} `json:"usage"`
			Content []struct {
				Text string `json:"text"`
			} `json:"content"`
		}
		if json.Unmarshal(body, &r) == nil {
			u.in = r.Usage.InputTokens
			u.out = r.Usage.OutputTokens
			u.cacheWrite = r.Usage.CacheCreationInputTokens
			u.cacheRead = r.Usage.CacheReadInputTokens
			var sb strings.Builder
			for _, c := range r.Content {
				sb.WriteString(c.Text)
			}
			preview = sb.String()
		}
	} else {
		var r struct {
			Usage struct {
				PromptTokens     int `json:"prompt_tokens"`
				CompletionTokens int `json:"completion_tokens"`
			} `json:"usage"`
			Choices []struct {
				Message struct {
					Content string `json:"content"`
				} `json:"message"`
			} `json:"choices"`
		}
		if json.Unmarshal(body, &r) == nil {
			u.in = r.Usage.PromptTokens
			u.out = r.Usage.CompletionTokens
			if len(r.Choices) > 0 {
				preview = r.Choices[0].Message.Content
			}
		}
	}
	return u, trim(preview, 400)
}

func trim(s string, n int) string {
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.TrimSpace(s)
	if len(s) > n {
		return s[:n] + "..."
	}
	return s
}

func (p *Proxy) ListenAddr() string {
	return fmt.Sprintf("127.0.0.1:%d", p.Port)
}

File: src/proxy/proxy.ts
import { Event, Bus } from "../event/bus.js";
import { Store } from "../store/store.js";
import { cost as pricingCost } from "../pricing/pricing.js";

interface Route {
  provider: "anthropic" | "openai";
  target: string;
}

interface RequestMeta {
  model: string;
  streaming: boolean;
  promptPreview: string;
}

interface Usage {
  in: number;
  out: number;
  cacheRead: number;
  cacheWrite: number;
}

// CaptureWriter is a 1MB ring buffer that retains the tail of a stream.
// The final SSE chunk (which contains the usage payload) arrives last, so
// we keep the last 1MB instead of the first 1MB.
class CaptureWriter {
  private buf: Buffer = Buffer.alloc(0);
  private readonly max = 1 << 20;

  write(chunk: Buffer): void {
    const n = chunk.length;
    if (this.buf.length + n <= this.max) {
      this.buf = Buffer.concat([this.buf, chunk]);
      return;
    }
    // Overflow: keep the last max bytes.
    const combined = Buffer.concat([this.buf, chunk]);
    if (combined.length > this.max) {
      this.buf = combined.subarray(combined.length - this.max);
    } else {
      this.buf = combined;
    }
  }

  bytes(): Buffer {
    return this.buf;
  }
}

export interface ProxyOptions {
  port: number;
  anthropicURL?: string;
  openaiURL?: string;
}

export class Proxy {
  private store: Store;
  private bus: Bus;
  private port: number;
  private anthropicTarget: string;
  private openaiTarget: string;
  private traceId = 0;

  constructor(store: Store, bus: Bus, opts: ProxyOptions) {
    this.store = store;
    this.bus = bus;
    this.port = opts.port;
    this.anthropicTarget = opts.anthropicURL || "https://api.anthropic.com";
    this.openaiTarget = opts.openaiURL || "https://api.openai.com";
  }

  listenAddr(): string {
    return `127.0.0.1:${this.port}`;
  }

  private route(req: Request, path: string): Route {
    // x-agenttop-upstream: explicit upstream override (used by opencode-go,
    // opencode-zen, and any custom provider that sets this header). This lets
    // the proxy forward to the real API even for providers it doesn't know.
    const upstream = req.headers.get("x-agenttop-upstream");
    if (upstream) {
      const isAnthropic =
        req.headers.get("anthropic-version") !== null ||
        path.startsWith("/v1/messages");
      return {
        provider: isAnthropic ? "anthropic" : "openai",
        target: upstream,
      };
    }
    const explicit = req.headers.get("x-agenttop-provider");
    if (explicit === "anthropic")
      return { provider: "anthropic", target: this.anthropicTarget };
    if (explicit === "openai")
      return { provider: "openai", target: this.openaiTarget };
    // The Anthropic SDK (used by opencode, Claude Code, etc.) sends
    // `anthropic-version` on every request. This is more robust than
    // path matching: it also catches /v1/messages/count_tokens and any
    // future endpoints, routing them to Anthropic instead of OpenAI.
    if (req.headers.get("anthropic-version") !== null)
      return { provider: "anthropic", target: this.anthropicTarget };
    if (path.startsWith("/v1/messages"))
      return { provider: "anthropic", target: this.anthropicTarget };
    return { provider: "openai", target: this.openaiTarget };
  }

  private async emit(e: Event): Promise<void> {
    const stored = await this.store.append(e);
    this.bus.emit(stored);
  }

  async handle(req: Request, path: string): Promise<Response> {
    const rt = this.route(req, path);
    const traceId = ++this.traceId;
    const start = new Date();

    let body: Buffer;
    try {
      const ab = await req.arrayBuffer();
      body = Buffer.from(ab);
    } catch (err) {
      return new Response(String(err), { status: 400 });
    }

    const meta = parseRequest(body);

    const startEvt: Event = {
      id: 0,
      traceId,
      time: start,
      provider: rt.provider,
      model: meta.model,
      endpoint: path,
      method: req.method,
      status: 0,
      streaming: meta.streaming,
      durationMs: 0,
      inputTokens: 0,
      outputTokens: 0,
      cacheReadTokens: 0,
      cacheWriteTokens: 0,
      costUSD: 0,
      promptPreview: meta.promptPreview,
      responsePreview: "",
      err: "",
    };
    await this.emit(startEvt);

    // Build upstream URL.
    const inUrl = new URL(req.url);
    const outURL = rt.target + inUrl.pathname + inUrl.search;

    // Copy headers, dropping Host/Content-Length and the agenttop hints.
    const outHeaders = new Headers();
    for (const [k, v] of req.headers.entries()) {
      const lower = k.toLowerCase();
      if (
        lower === "host" ||
        lower === "content-length" ||
        lower === "x-agenttop-provider" ||
        lower === "x-agenttop-upstream"
      ) {
        continue;
      }
      outHeaders.set(k, v);
    }

    let upstream: Response;
    try {
      upstream = await fetch(outURL, {
        method: req.method,
        headers: outHeaders,
        body: req.method === "GET" || req.method === "HEAD" ? undefined : body,
      });
    } catch (err) {
      const fail = makeFail(startEvt, start, String(err));
      await this.emit(fail);
      return new Response(String(err), { status: 502 });
    }

    // Build response headers (drop content-length since we're streaming
    // a different body).
    const respHeaders = new Headers();
    for (const [k, v] of upstream.headers.entries()) {
      if (k.toLowerCase() === "content-length") continue;
      respHeaders.set(k, v);
    }

    const ct = upstream.headers.get("content-type") || "";
    const isSSE = ct.includes("text/event-stream");
    const capture = new CaptureWriter();

    // If there's no body (e.g. HEAD, 204), return immediately.
    if (!upstream.body) {
      const endEvt = makeEndEvt(startEvt, start, upstream.status, { in: 0, out: 0, cacheRead: 0, cacheWrite: 0 }, "", meta.model);
      await this.emit(endEvt);
      return new Response(null, { status: upstream.status, headers: respHeaders });
    }

    // Stream the upstream body to the client in real-time while
    // capturing chunks in parallel. When the stream ends, parse the
    // captured tail for usage and emit the end event.
    const reader = upstream.body.getReader();

    const stream = new ReadableStream({
      pull: async (controller): Promise<void> => {
        const { done, value } = await reader.read();
        if (done) {
          controller.close();
          // Parse usage from the captured tail.
          const { usage, responsePreview } = isSSE
            ? parseSSE(rt.provider, capture.bytes())
            : parseJSON(rt.provider, capture.bytes());
          const endEvt = makeEndEvt(startEvt, start, upstream.status, usage, responsePreview, meta.model);
          await this.emit(endEvt);
          return;
        }
        if (value) {
          capture.write(Buffer.from(value));
          controller.enqueue(value);
        }
      },
      cancel: () => {
        reader.cancel().catch(() => {});
      },
    });

    return new Response(stream, {
      status: upstream.status,
      headers: respHeaders,
    });
  }
}

function makeFail(e: Event, start: Date, msg: string): Event {
  return {
    ...e,
    time: new Date(),
    status: 0,
    durationMs: Date.now() - start.getTime(),
    err: msg,
  };
}

function makeEndEvt(
  startEvt: Event,
  start: Date,
  status: number,
  usage: Usage,
  responsePreview: string,
  model: string,
): Event {
  return {
    ...startEvt,
    time: new Date(),
    status,
    durationMs: Date.now() - start.getTime(),
    inputTokens: usage.in,
    outputTokens: usage.out,
    cacheReadTokens: usage.cacheRead,
    cacheWriteTokens: usage.cacheWrite,
    responsePreview,
    costUSD: pricingCost(model, usage.in, usage.out, usage.cacheRead, usage.cacheWrite),
  };
}

function trim(s: string, n: number): string {
  const oneLine = s.replace(/\s+/g, " ").trim();
  if (oneLine.length > n) return oneLine.slice(0, n) + "…";
  return oneLine;
}

function parseRequest(body: Buffer): RequestMeta {
  try {
    const j = JSON.parse(body.toString("utf-8")) as {
      model?: string;
      stream?: boolean;
      messages?: Array<{ role: string; content: unknown }>;
    };
    let preview = "";
    if (Array.isArray(j.messages)) {
      for (let i = j.messages.length - 1; i >= 0; i--) {
        if (j.messages[i].role?.toLowerCase() === "user") {
          preview = extractText(j.messages[i].content);
          break;
        }
      }
    }
    return {
      model: j.model || "",
      streaming: Boolean(j.stream),
      promptPreview: trim(preview, 120),
    };
  } catch {
    return { model: "", streaming: false, promptPreview: "" };
  }
}

function extractText(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((p) => {
        if (typeof p === "object" && p !== null) {
          const pp = p as { type?: string; text?: string };
          if (!pp.type || pp.type === "text") return pp.text || "";
        }
        return "";
      })
      .join("");
  }
  return "";
}

function parseSSE(
  provider: "anthropic" | "openai",
  body: Buffer,
): { usage: Usage; responsePreview: string } {
  const u: Usage = { in: 0, out: 0, cacheRead: 0, cacheWrite: 0 };
  const text: string[] = [];

  for (const rawLine of body.toString("utf-8").split("\n")) {
    const line = rawLine.trim();
    if (!line.startsWith("data:")) continue;
    const data = line.slice(5).trim();
    if (data === "[DONE]") continue;
    let obj: Record<string, unknown>;
    try {
      obj = JSON.parse(data) as Record<string, unknown>;
    } catch {
      continue;
    }
    if (provider === "anthropic") {
      const t = obj.type as string | undefined;
      if (t === "message_start") {
        const wrap = obj as {
          message?: { usage?: Record<string, number> };
        };
        const usage = wrap.message?.usage;
        if (usage) {
          u.in = usage.input_tokens || 0;
          u.cacheWrite = usage.cache_creation_input_tokens || 0;
          u.cacheRead = usage.cache_read_input_tokens || 0;
        }
      } else if (t === "message_delta") {
        const wrap = obj as { usage?: { output_tokens?: number } };
        const out = wrap.usage?.output_tokens;
        if (out && out > 0) u.out = out;
      }
      // Extract text deltas
      const d = obj.delta as { type?: string; text?: string } | undefined;
      if (d?.type === "text_delta" && d.text) text.push(d.text);
    } else {
      const usage = obj.usage as
        | {
            prompt_tokens?: number;
            completion_tokens?: number;
            prompt_cache_hit_tokens?: number;
          }
        | null;
      if (usage) {
        if (usage.prompt_tokens && usage.prompt_tokens > 0)
          u.in = usage.prompt_tokens;
        if (usage.prompt_cache_hit_tokens && usage.prompt_cache_hit_tokens > 0)
          u.cacheRead = usage.prompt_cache_hit_tokens;
        if (usage.completion_tokens && usage.completion_tokens > 0)
          u.out = usage.completion_tokens;
      }
      const choices = obj.choices as
        | Array<{ delta?: { content?: string; reasoning_content?: string } }>
        | undefined;
      if (Array.isArray(choices)) {
        for (const c of choices) {
          if (c.delta?.content) text.push(c.delta.content);
        }
      }
    }
  }
  return { usage: u, responsePreview: trim(text.join(""), 400) };
}

function parseJSON(
  provider: "anthropic" | "openai",
  body: Buffer,
): { usage: Usage; responsePreview: string } {
  const u: Usage = { in: 0, out: 0, cacheRead: 0, cacheWrite: 0 };
  let preview = "";
  try {
    const j = JSON.parse(body.toString("utf-8")) as Record<string, unknown>;
    if (provider === "anthropic") {
      const usage = j.usage as
        | {
            input_tokens?: number;
            output_tokens?: number;
            cache_creation_input_tokens?: number;
            cache_read_input_tokens?: number;
          }
        | undefined;
      if (usage) {
        u.in = usage.input_tokens || 0;
        u.out = usage.output_tokens || 0;
        u.cacheWrite = usage.cache_creation_input_tokens || 0;
        u.cacheRead = usage.cache_read_input_tokens || 0;
      }
      const content = j.content as Array<{ text?: string }> | undefined;
      if (Array.isArray(content)) {
        preview = content.map((c) => c.text || "").join("");
      }
    } else {
      const usage = j.usage as
        | { prompt_tokens?: number; completion_tokens?: number }
        | undefined;
      if (usage) {
        u.in = usage.prompt_tokens || 0;
        u.out = usage.completion_tokens || 0;
      }
      const choices = j.choices as
        | Array<{ message?: { content?: string } }>
        | undefined;
      if (Array.isArray(choices) && choices[0]?.message) {
        preview = choices[0].message.content || "";
      }
    }
  } catch {
    // Not JSON, leave usage at zero.
  }
  return { usage: u, responsePreview: trim(preview, 400) };
}

// startProxy starts a Bun HTTP server that routes all traffic through
// the proxy logic. Returns the actual port (useful when the caller
// passed 0 for "pick any free port").
export async function startProxy(
  proxy: Proxy,
  port: number,
): Promise<{ server: ReturnType<typeof Bun.serve>; port: number }> {
  const server = Bun.serve({
    port,
    hostname: "127.0.0.1",
    async fetch(req) {
      const url = new URL(req.url);
      return proxy.handle(req, url.pathname);
    },
  });
  return { server, port: server.port ?? port };
}
